### ISSUE TITLE
improve multi schema handling in different auth flows

### ISSUE THREAD

**Issue by c0depwn** [ory/kratos#765]
**Is your feature request related to a problem? Please describe.**

I came across an issue when spinning up kratos along with the oathkeeper proxy and selfservice-ui-node. I have defined multiple identity schemas in json inside the kratos config. One is for an “employee” user type and the second one for a “customer” user type. My issue is that I would like to be able to specify the schema when registering a new account from the selfservice endpoints or public API or anywhere else where I can create/edit a user. 
The admin API seems to offer a `POST /identities` which allows you to send `schema_id` in the request body which works, but there currently is no way to specify any credentials in this way (which makes sense and I think is already addressed in v0.5). Also when I create a new user via the selfservice-ui-node and check the created identity in the admin API with `GET /identities` the `schema_id` always seems to be “default” since I cannot specify one in the selfservice flows. Only when using the previously mentioned admin API endpoint to create a user the schema is actually “employee” or “customer” instead of “default”.

It would be awesome to have some way to distinguish between schemas and to use identity `schema_id`'s similar to how a "role" would work. Especially when having a SPA, it would be awesome to be able to restrict the UI according to the `schema_id` which is retrieved inside the response of the `GET /session/whoami` or in the JWT sub on the backend when using oathkeeper along with kratos. Additionally, I think it would be an improvement to actually set the `schema_id` instead of "default" when creating a user with the current implementation through the selfservice registration flow.

**Describe the solution you'd like**

Specify `schema_id` in flows effecting identity creation/modification in some way to render proper traits/form inputs. Additionally, store the proper `schema_id` instead of "default", even when only 1 schema is specified in the config.

**Describe alternatives you've considered**

None at the moment.

**Additional context**

This feature brings along some security concerns, since we wouldn't want to be able to let a user register with, let's say, an "admin" schema, without actually having the right to do so.

**Comment by aeneasr** [comment]
Thank you for the write-up and sorry for the late response!

> Specify schema_id in flows effecting identity creation/modification in some way to render proper traits/form inputs. Additionally, store the proper schema_id instead of "default", even when only 1 schema is specified in the config.

The problem here is that the user is allowed to pick the schema to their liking. Are they really an employee? Or do they just say they're one? When it comes to roles in a company (which the schema in this case would infer) it is very dangerous to leave that open.

Generally though, I think it should be possible to set the schema for registration flows to something other than default. Would that maybe address your issue?

Alternatively, we could probably offer an Admin API endpoint which initializes a new registration flow with the given schema ID. Then you would have a privileged backend service initializing the flow and setting the flow's schema ID to the value you want.

**Comment by c0depwn** [comment]
I think it would make sense to modify the Admin API so that an authorised backend could choose a schema by supplying the `schema_id` when initialising the registration flow.

**Comment by aeneasr** [comment]
Yup, that sounds the most reasonable to me as well!

**Comment by atreya2011** [comment]
@aeneasr 
First of all, thank you for providing such an excellent service as open source.

> Generally though, I think it should be possible to set the schema for registration flows to something other than default. Would that maybe address your issue?
Please correct me if I am wrong. By doing the above, we should technically be able to specify the schema for self-service registration flows right?

If that is the case, then does this endpoint requires modification?
https://www.ory.sh/kratos/docs/next/reference/api#operation/initializeSelfServiceRegistrationFlowForBrowsers

For example, I would like to for example implement a customer registration UI that automatically saves the customer information using the customer json schema.

**Comment by gitn00b1337** [comment]
I'd argue this should be "both". 
In this issues example then the employee is a restricted account type you'd want to hide. But for many SAAS products there's multiple account types (business customer, end users, affiliates etc) which all would be fine to have public registration for. And if there's specific logic required to "validate" the account registration before acceptance you can already do that via hooks.

**Comment by Kaishu-Sahu** [comment]
I also think there should be both because in our use case, we have two types of customers phone/password and email/password. We want to keep them separated as both have different roles to do.

**Comment by nikist97** [comment]
I am facing the same problem. Two different account types (so different schemas) - for simplicity, I would just call one of the account types a "business owner account" and the other "a customer account". Users are able to register as a regular customer or as a business owner.

I was trying to find in the API spec if there is a way to initiate the registration flow with a non-default schema ID, but this doesn't seem to be possbile based on the information in this issue.

Any advice?

**Comment by ramirobg94** [comment]
I am also trying to understand what is the best way to implement different roles, any idea or documentation?

Also I need to populate the role into the claims to be used by other services like hasura

**Comment by sijumoncy** [comment]
Any updation in  passing schemas id in registration flow instead of default schema . ?

**Comment by aeneasr** [comment]
Yes, the design should be like this. When specifying identity schemas, one should be able to specify whether a schema can be used for registration (naming is just a suggestion):

```
identity:
  schemas:
    - id: ...
      url: ...
      self_service_available: true
```

Then, during the init registration flow, it can be set as a query parameter:

```
GET /self-service/.../browser?use_schema={schema_id}
```

**Comment by nikist97** [comment]
@aeneasr do you know if anyone is working on implementing this suggestion?

**Comment by aeneasr** [comment]
Afaik not!

**Comment by nikist97** [comment]
@aeneasr Would you say this is something a first-time contributor can implement? I haven't touched the Kratos code base before, but will be willing to spend some time on this since it's important for the project I am working on

**Comment by aeneasr** [comment]
Depends a bit on how well you know Go but if you do I think it should be doable. The most difficult part will be writing the tests most likely!

**Comment by nikist97** [comment]
Got it, I will give it a try, hope you don't mind me reaching out on Slack if I get stuck

**Comment by aeneasr** [comment]
Of course :)

**Comment by nikist97** [comment]
@sijumoncy my understanding is that the suggestion from @aeneasr (I will let him correct me if I am wrong) will work the same way for both browser-based and API-based flows - a query parameter `use_schema` so in your case `self-service/registration/api?use_schema=<schema_id>`

**Comment by MJEND7** [comment]
@nikist97 Did you end up starting on this? This is something that I would love to have working!

**Comment by nikist97** [comment]
@MJEND7 I have briefly looked through the code, but unfortunately didn't have the chance to actively work on it yet due to project re-prioritization.  I was hoping to find the time to work on it this month, but if you have an urgent need feel free to pick it up instead of me

**Comment by nikist97** [comment]
Just providing a quick update here, I finally found some time to try to implement this feature and have an initial version; currently working on fixing existing tests and adding new ones

@aeneasr hope you don't mind if I reach out here or on Slack in case I get stuck

**Comment by nikist97** [comment]
Unfortunately, this seems like a more complex feature than I originally expected since there's lots of logic in the code that assumes the default schema was used during registration. My suspicion is that the registration flow instance will need to be updated to store the chosen schema ID in order to be able to implement this (pending discussion with @aeneasr)

**Comment by till** [comment]
Is there a document to read on what the schemas can be used for? Aside from the admin APIs, they don't seem to be used in any APIs? And what's the correlation with identifiers, can I have multiple identities that share the same email address, but use different schemas?

**Comment by zeroows** [comment]
Has progress been made by having different schemas work with login/registration flows?

**Comment by acmayberry** [comment]
I'm also curious about progress on this issue.

**Comment by renom** [comment]
What if, let's say I need two different registration (login) flows:
1. The first flow accepts one schema (registration and login from internet).
2. The second flow accepts multiple schemas (registration and login from internal network).

I.e. I allow the first case flows from internet with oathkeeper, while the second case flows are only allowed from internal network.
Can I achieve this?
If no, please consider supporting it.

**Comment by renom** [comment]
Can I run a second instance of kratos with a different config (but the same db) as a workaround?
